File: source/FAST/DataStream.cpp
#include "DataStream.hpp"
#include "ProcessObject.hpp"

namespace fast {

bool DataStream::isDone() {
    return m_done;
}

DataStream::DataStream(std::vector<std::shared_ptr<ProcessObject>> processObjects) {
    m_processObjects = processObjects;
    for(auto po : m_processObjects) {
        for(int i = 0; i < po->getNrOfOutputPorts(); ++i) {
            m_outputPorts.push_back(po->getOutputPort(i));
        }
    }
    if(m_outputPorts.empty())
        throw Exception("Process objects given to DataStream had no output ports");
}

std::shared_ptr<DataObject> DataStream::getNextDataObject(uint portID) {
    if(m_outputPorts.size() <= portID)
        throw Exception("Output nr " + std::to_string(portID) + " does not exist in DataStream.");
    if(m_nextDataObjects.count(portID) == 0) {
        // Run all POs
        for(int i = 0; i < m_outputPorts.size(); ++i) {
            m_outputPorts[i]->getProcessObject()->run(m_executeToken);
            m_nextDataObjects[i] = m_outputPorts[i]->getNextFrame();
            if(m_nextDataObjects[i]->isLastFrame()) {
                m_done = true;
            }
        }
        ++m_executeToken;
    }
    auto result = m_nextDataObjects[portID];
    m_nextDataObjects.erase(portID);
    return std::shared_ptr<DataObject>(result);
}

DataStream::DataStream(std::shared_ptr<ProcessObject> po) : DataStream(std::vector<std::shared_ptr<ProcessObject>>({po})) {

}

}

File: source/FAST/Python/DataStream.py
class DataStream:
    def __init__(self, *args):
        self._processObjects = args;
        self._ports = []
        self.finished = False
        self._executeToken = 0
        for po in self._processObjects:
            for i in range(po.getNrOfOutputPorts()):
                self._ports.append(po.getOutputPort(i))
        if len(self._ports) == 0:
            raise Exception('Can\'t create data stream (getDataStream) on a process object with 0 output ports')
    def __iter__(self):
        return self
    def __next__(self):
        if self.finished:
            raise StopIteration

        for port in self._ports:
            port.getProcessObject().run(self._executeToken)
        self._executeToken += 1
        if self._executeToken < 0:
            self._executeToken = 0
        result = []
        for port in self._ports:
            data = port.getNextFrame()
            if data.isLastFrame(): self.finished = True
            result.append(data)
        if len(result) == 1:
            return result[0]
        else:
            return result
